### DISCUSSION TITLE
Interesting performance characteristics on the Nvidia Jetson

### DISCUSSION THREAD

**Discussion by arduano** [ogxd/gxhash#41] [General]
Just thought I'd share...

![image](https://github.com/ogxd/gxhash/assets/13347712/3b936f95-9161-4b6a-9ced-4f86b0d8640e)

Ran the tests multiple times to make sure this was not an accident.

**Comment by ogxd** [comment]
That's indeed very interesting! 128 bytes is the size at which the algorithm starts to proceed with an "high ILP construction" for greater throughput. It seems to underperform on this hardware compared to the <128 bytes processing, I'd be curious to understand why but it might be difficult to put my hands on such hardware.  
I'm not familiar with this product. Are you using the "[Jetson Orin Nano Developer Kit](https://developer.nvidia.com/embedded/jetson-agx-orin-developer-kit)"? It seems it uses an ARM Cortex-A78AE CPU, but the throughput does not depend on the CPU alone but also on memory bandwidth.

**Reply by arduano** [reply]
It's a Xavier NX, or more specifically, `NVIDIA Jetson Xavier NX Developer Kit`.

**Reply by arduano** [reply]
I wouldn't mind running more tests btw, especially if it just involves checking out to a branch and running the command. I'm the only person using the jetson at the office at the moment.

**Reply by ogxd** [reply]
Can you checkout `hybrid-aesmacros-exp` and run the bench again? This is the candidate branch for the next major version. There has been significant changes.

**Reply by arduano** [reply]
One moment, I think it would be interesting to try the low power mode as well, the original benchmark was on 20W 6-core mode, so I'm trying 10W 2-core mode right now (currently running)

**Reply by arduano** [reply]
![image](https://github.com/ogxd/gxhash/assets/13347712/5cee12c7-3596-4db4-90be-fd867e1cc789)
10W 2-core, main branch

**Reply by arduano** [reply]
/proc/cpuinfo reports identical CPU features on each core in both 10W and 20W modes, but seems like 64 bit gets hit in the above graph

**Reply by arduano** [reply]
![image](https://github.com/ogxd/gxhash/assets/13347712/654f24e9-690a-4c58-8441-503809519a20)
20W 6-core `hybrid-aesmacros-exp` branch

**Reply by arduano** [reply]
Honestly that looks a bit suspicious, I'll try again

**Reply by ogxd** [reply]
Maybe there is some benchmark biais going on? Can you try `cargo bench --bench throughput_criterion` ?

**Reply by arduano** [reply]
Give me a bit, just re-compiling everything again (part of me being suspicious was that running again didn't trigger a cargo compile after switching branches, it just ran)

**Reply by arduano** [reply]
![image](https://github.com/ogxd/gxhash/assets/13347712/0f1d2524-03c3-438e-9973-486b8df98e37)
Previous command, same environment (not the new command)
Running new command rn

**Reply by arduano** [reply]
This one is taking much longer, but here's results from gxhash:
```
all/gxhash/4            time:   [169.82 ps 170.37 ps 171.01 ps]
                        thrpt:  [21.784 GiB/s 21.866 GiB/s 21.937 GiB/s]
Found 94 outliers among 1000 measurements (9.40%)
  32 (3.20%) high mild
  62 (6.20%) high severe
all/gxhash/16           time:   [172.71 ps 174.08 ps 175.64 ps]
                        thrpt:  [84.838 GiB/s 85.601 GiB/s 86.278 GiB/s]
Found 122 outliers among 1000 measurements (12.20%)
  29 (2.90%) high mild
  93 (9.30%) high severe
all/gxhash/64           time:   [169.12 ps 169.49 ps 169.90 ps]
                        thrpt:  [350.83 GiB/s 351.68 GiB/s 352.43 GiB/s]
Found 134 outliers among 1000 measurements (13.40%)
  44 (4.40%) high mild
  90 (9.00%) high severe
all/gxhash/256          time:   [15.844 ns 15.997 ns 16.165 ns]
                        thrpt:  [14.749 GiB/s 14.904 GiB/s 15.047 GiB/s]
Found 109 outliers among 1000 measurements (10.90%)
  24 (2.40%) high mild
  85 (8.50%) high severe
all/gxhash/1024         time:   [75.976 ns 76.551 ns 77.201 ns]
                        thrpt:  [12.353 GiB/s 12.458 GiB/s 12.552 GiB/s]
Found 105 outliers among 1000 measurements (10.50%)
  35 (3.50%) high mild
  70 (7.00%) high severe
all/gxhash/4096         time:   [256.68 ns 258.73 ns 260.97 ns]
                        thrpt:  [14.617 GiB/s 14.744 GiB/s 14.862 GiB/s]
Found 108 outliers among 1000 measurements (10.80%)
  22 (2.20%) high mild
  86 (8.60%) high severe
all/gxhash/16384        time:   [957.11 ns 962.85 ns 969.09 ns]
                        thrpt:  [15.745 GiB/s 15.848 GiB/s 15.942 GiB/s]
Found 121 outliers among 1000 measurements (12.10%)
  33 (3.30%) high mild
  88 (8.80%) high severe
all/gxhash/65536        time:   [3.8259 µs 3.8435 µs 3.8628 µs]
                        thrpt:  [15.801 GiB/s 15.880 GiB/s 15.953 GiB/s]
Found 145 outliers among 1000 measurements (14.50%)
  37 (3.70%) high mild
  108 (10.80%) high severe
all/gxhash/262144       time:   [14.693 µs 14.728 µs 14.774 µs]
                        thrpt:  [16.525 GiB/s 16.577 GiB/s 16.616 GiB/s]
```

The other hash algorithms are still computing, will show the criterion graphs when they're done.

**Reply by arduano** [reply]
<details>
  <summary>Command output</summary>
<pre>
all/gxhash/4            time:   [169.82 ps 170.37 ps 171.01 ps]
                        thrpt:  [21.784 GiB/s 21.866 GiB/s 21.937 GiB/s]
Found 94 outliers among 1000 measurements (9.40%)
  32 (3.20%) high mild
  62 (6.20%) high severe
all/gxhash/16           time:   [172.71 ps 174.08 ps 175.64 ps]
                        thrpt:  [84.838 GiB/s 85.601 GiB/s 86.278 GiB/s]
Found 122 outliers among 1000 measurements (12.20%)
  29 (2.90%) high mild
  93 (9.30%) high severe
all/gxhash/64           time:   [169.12 ps 169.49 ps 169.90 ps]
                        thrpt:  [350.83 GiB/s 351.68 GiB/s 352.43 GiB/s]
Found 134 outliers among 1000 measurements (13.40%)
  44 (4.40%) high mild
  90 (9.00%) high severe
all/gxhash/256          time:   [15.844 ns 15.997 ns 16.165 ns]
                        thrpt:  [14.749 GiB/s 14.904 GiB/s 15.047 GiB/s]
Found 109 outliers among 1000 measurements (10.90%)
  24 (2.40%) high mild
  85 (8.50%) high severe
all/gxhash/1024         time:   [75.976 ns 76.551 ns 77.201 ns]
                        thrpt:  [12.353 GiB/s 12.458 GiB/s 12.552 GiB/s]
Found 105 outliers among 1000 measurements (10.50%)
  35 (3.50%) high mild
  70 (7.00%) high severe
all/gxhash/4096         time:   [256.68 ns 258.73 ns 260.97 ns]
                        thrpt:  [14.617 GiB/s 14.744 GiB/s 14.862 GiB/s]
Found 108 outliers among 1000 measurements (10.80%)
  22 (2.20%) high mild
  86 (8.60%) high severe
all/gxhash/16384        time:   [957.11 ns 962.85 ns 969.09 ns]
                        thrpt:  [15.745 GiB/s 15.848 GiB/s 15.942 GiB/s]
Found 121 outliers among 1000 measurements (12.10%)
  33 (3.30%) high mild
  88 (8.80%) high severe
all/gxhash/65536        time:   [3.8259 µs 3.8435 µs 3.8628 µs]
                        thrpt:  [15.801 GiB/s 15.880 GiB/s 15.953 GiB/s]
Found 145 outliers among 1000 measurements (14.50%)
  37 (3.70%) high mild
  108 (10.80%) high severe
all/gxhash/262144       time:   [14.693 µs 14.728 µs 14.774 µs]
                        thrpt:  [16.525 GiB/s 16.577 GiB/s 16.616 GiB/s]
Found 19 outliers among 1000 measurements (1.90%)
  11 (1.10%) high mild
  8 (0.80%) high severe
all/ahash/4             time:   [1.7699 ns 1.7890 ns 1.8099 ns]
                        thrpt:  [2.0582 GiB/s 2.0824 GiB/s 2.1048 GiB/s]
Found 99 outliers among 1000 measurements (9.90%)
  21 (2.10%) high mild
  78 (7.80%) high severe
all/ahash/16            time:   [2.1058 ns 2.1452 ns 2.1889 ns]
                        thrpt:  [6.8077 GiB/s 6.9462 GiB/s 7.0761 GiB/s]
Found 116 outliers among 1000 measurements (11.60%)
  24 (2.40%) high mild
  92 (9.20%) high severe
all/ahash/64            time:   [8.1441 ns 8.2399 ns 8.3443 ns]
                        thrpt:  [7.1432 GiB/s 7.2337 GiB/s 7.3187 GiB/s]
Found 109 outliers among 1000 measurements (10.90%)
  25 (2.50%) high mild
  84 (8.40%) high severe
all/ahash/256           time:   [42.895 ns 43.076 ns 43.288 ns]
                        thrpt:  [5.5078 GiB/s 5.5348 GiB/s 5.5582 GiB/s]
Found 80 outliers among 1000 measurements (8.00%)
  20 (2.00%) high mild
  60 (6.00%) high severe
all/ahash/1024          time:   [148.07 ns 149.53 ns 151.14 ns]
                        thrpt:  [6.3099 GiB/s 6.3777 GiB/s 6.4406 GiB/s]
Found 221 outliers among 1000 measurements (22.10%)
  13 (1.30%) high mild
  208 (20.80%) high severe
all/ahash/4096          time:   [540.04 ns 544.81 ns 550.09 ns]
                        thrpt:  [6.9347 GiB/s 7.0019 GiB/s 7.0638 GiB/s]
Found 91 outliers among 1000 measurements (9.10%)
  23 (2.30%) high mild
  68 (6.80%) high severe
all/ahash/16384         time:   [2.1114 µs 2.1185 µs 2.1268 µs]
                        thrpt:  [7.1747 GiB/s 7.2026 GiB/s 7.2268 GiB/s]
Found 89 outliers among 1000 measurements (8.90%)
  30 (3.00%) high mild
  59 (5.90%) high severe
all/ahash/65536         time:   [8.8342 µs 8.9311 µs 9.0367 µs]
                        thrpt:  [6.7542 GiB/s 6.8340 GiB/s 6.9089 GiB/s]
Found 90 outliers among 1000 measurements (9.00%)
  23 (2.30%) high mild
  67 (6.70%) high severe
all/ahash/262144        time:   [33.756 µs 33.887 µs 34.032 µs]
                        thrpt:  [7.1738 GiB/s 7.2045 GiB/s 7.2325 GiB/s]
Found 90 outliers among 1000 measurements (9.00%)
  25 (2.50%) high mild
  65 (6.50%) high severe
all/t1ha0/4             time:   [1.6304 ns 1.6391 ns 1.6492 ns]
                        thrpt:  [2.2588 GiB/s 2.2727 GiB/s 2.2849 GiB/s]
Found 83 outliers among 1000 measurements (8.30%)
  26 (2.60%) high mild
  57 (5.70%) high severe
all/t1ha0/16            time:   [2.3985 ns 2.4298 ns 2.4655 ns]
                        thrpt:  [6.0439 GiB/s 6.1327 GiB/s 6.2128 GiB/s]
Found 79 outliers among 1000 measurements (7.90%)
  27 (2.70%) high mild
  52 (5.20%) high severe
all/t1ha0/64            time:   [4.5172 ns 4.5249 ns 4.5348 ns]
                        thrpt:  [13.144 GiB/s 13.173 GiB/s 13.195 GiB/s]
Found 49 outliers among 1000 measurements (4.90%)
  15 (1.50%) high mild
  34 (3.40%) high severe
all/t1ha0/256           time:   [49.195 ns 49.511 ns 49.865 ns]
                        thrpt:  [4.7812 GiB/s 4.8155 GiB/s 4.8464 GiB/s]
Found 90 outliers among 1000 measurements (9.00%)
  29 (2.90%) high mild
  61 (6.10%) high severe
all/t1ha0/1024          time:   [152.14 ns 152.90 ns 153.77 ns]
                        thrpt:  [6.2021 GiB/s 6.2371 GiB/s 6.2684 GiB/s]
Found 100 outliers among 1000 measurements (10.00%)
  33 (3.30%) high mild
  67 (6.70%) high severe
all/t1ha0/4096          time:   [565.70 ns 570.41 ns 575.81 ns]
                        thrpt:  [6.6249 GiB/s 6.6877 GiB/s 6.7433 GiB/s]
Found 110 outliers among 1000 measurements (11.00%)
  33 (3.30%) high mild
  77 (7.70%) high severe
all/t1ha0/16384         time:   [2.2326 µs 2.2527 µs 2.2781 µs]
                        thrpt:  [6.6981 GiB/s 6.7736 GiB/s 6.8346 GiB/s]
Found 146 outliers among 1000 measurements (14.60%)
  25 (2.50%) high mild
  121 (12.10%) high severe
all/t1ha0/65536         time:   [9.1037 µs 9.1583 µs 9.2192 µs]
                        thrpt:  [6.6204 GiB/s 6.6645 GiB/s 6.7045 GiB/s]
Found 120 outliers among 1000 measurements (12.00%)
  23 (2.30%) high mild
  97 (9.70%) high severe
all/t1ha0/262144        time:   [36.133 µs 37.105 µs 38.313 µs]
                        thrpt:  [6.3723 GiB/s 6.5797 GiB/s 6.7567 GiB/s]
Found 112 outliers among 1000 measurements (11.20%)
  44 (4.40%) high mild
  68 (6.80%) high severe
all/xxhash/4            time:   [169.05 ps 169.24 ps 169.46 ps]
                        thrpt:  [21.984 GiB/s 22.012 GiB/s 22.036 GiB/s]
Found 61 outliers among 1000 measurements (6.10%)
  17 (1.70%) high mild
  44 (4.40%) high severe
all/xxhash/16           time:   [852.72 ps 861.16 ps 870.77 ps]
                        thrpt:  [17.113 GiB/s 17.304 GiB/s 17.475 GiB/s]
Found 76 outliers among 1000 measurements (7.60%)
  21 (2.10%) high mild
  55 (5.50%) high severe
all/xxhash/64           time:   [2.7452 ns 2.7588 ns 2.7738 ns]
                        thrpt:  [21.489 GiB/s 21.605 GiB/s 21.712 GiB/s]
Found 124 outliers among 1000 measurements (12.40%)
  32 (3.20%) high mild
  92 (9.20%) high severe
all/xxhash/256          time:   [59.666 ns 60.203 ns 60.798 ns]
                        thrpt:  [3.9215 GiB/s 3.9603 GiB/s 3.9959 GiB/s]
Found 141 outliers among 1000 measurements (14.10%)
  27 (2.70%) high mild
  114 (11.40%) high severe
all/xxhash/1024         time:   [169.47 ns 171.16 ns 173.00 ns]
                        thrpt:  [5.5124 GiB/s 5.5719 GiB/s 5.6273 GiB/s]
Found 114 outliers among 1000 measurements (11.40%)
  31 (3.10%) high mild
  83 (8.30%) high severe
all/xxhash/4096         time:   [612.87 ns 615.55 ns 618.58 ns]
                        thrpt:  [6.1668 GiB/s 6.1972 GiB/s 6.2243 GiB/s]
Found 110 outliers among 1000 measurements (11.00%)
  36 (3.60%) high mild
  74 (7.40%) high severe
all/xxhash/16384        time:   [2.3925 µs 2.4033 µs 2.4182 µs]
                        thrpt:  [6.3100 GiB/s 6.3490 GiB/s 6.3776 GiB/s]
Found 103 outliers among 1000 measurements (10.30%)
  28 (2.80%) high mild
  75 (7.50%) high severe
all/xxhash/65536        time:   [9.8658 µs 9.9346 µs 10.011 µs]
                        thrpt:  [6.0968 GiB/s 6.1437 GiB/s 6.1865 GiB/s]
Found 73 outliers among 1000 measurements (7.30%)
  16 (1.60%) high mild
  57 (5.70%) high severe
all/xxhash/262144       time:   [37.794 µs 38.039 µs 38.315 µs]
                        thrpt:  [6.3720 GiB/s 6.4182 GiB/s 6.4598 GiB/s]
Found 67 outliers among 1000 measurements (6.70%)
  22 (2.20%) high mild
  45 (4.50%) high severe
all/highwayhash/4       time:   [78.318 ns 78.575 ns 78.869 ns]
                        thrpt:  [48.367 MiB/s 48.548 MiB/s 48.708 MiB/s]
Found 101 outliers among 1000 measurements (10.10%)
  43 (4.30%) high mild
  58 (5.80%) high severe
all/highwayhash/16      time:   [84.390 ns 84.659 ns 84.988 ns]
                        thrpt:  [179.54 MiB/s 180.24 MiB/s 180.81 MiB/s]
Found 26 outliers among 1000 measurements (2.60%)
  14 (1.40%) high mild
  12 (1.20%) high severe
all/highwayhash/64      time:   [66.839 ns 67.262 ns 67.738 ns]
                        thrpt:  [901.05 MiB/s 907.43 MiB/s 913.17 MiB/s]
Found 88 outliers among 1000 measurements (8.80%)
  21 (2.10%) high mild
  67 (6.70%) high severe
all/highwayhash/256     time:   [140.51 ns 141.35 ns 142.29 ns]
                        thrpt:  [1.6756 GiB/s 1.6868 GiB/s 1.6968 GiB/s]
Found 84 outliers among 1000 measurements (8.40%)
  19 (1.90%) high mild
  65 (6.50%) high severe
all/highwayhash/1024    time:   [384.03 ns 384.83 ns 385.75 ns]
                        thrpt:  [2.4722 GiB/s 2.4782 GiB/s 2.4833 GiB/s]
Found 109 outliers among 1000 measurements (10.90%)
  33 (3.30%) high mild
  76 (7.60%) high severe
all/highwayhash/4096    time:   [1.3731 µs 1.3772 µs 1.3820 µs]
                        thrpt:  [2.7603 GiB/s 2.7699 GiB/s 2.7782 GiB/s]
Found 81 outliers among 1000 measurements (8.10%)
  26 (2.60%) high mild
  55 (5.50%) high severe
Benchmarking all/highwayhash/16384: Warming up for 3.0000 s
Warning: Unable to complete 1000 samples in 2.0s. You may wish to increase target time to 2.7s, enable flat sampling, or reduce sample count to 610.
all/highwayhash/16384   time:   [5.2767 µs 5.2876 µs 5.3003 µs]
                        thrpt:  [2.8789 GiB/s 2.8858 GiB/s 2.8918 GiB/s]
Found 112 outliers among 1000 measurements (11.20%)
  50 (5.00%) high mild
  62 (6.20%) high severe
all/highwayhash/65536   time:   [21.379 µs 21.436 µs 21.501 µs]
                        thrpt:  [2.8387 GiB/s 2.8473 GiB/s 2.8549 GiB/s]
Found 50 outliers among 1000 measurements (5.00%)
  13 (1.30%) high mild
  37 (3.70%) high severe
all/highwayhash/262144  time:   [84.015 µs 84.361 µs 84.747 µs]
                        thrpt:  [2.8808 GiB/s 2.8940 GiB/s 2.9059 GiB/s]
Found 68 outliers among 1000 measurements (6.80%)
  19 (1.90%) high mild
  49 (4.90%) high severe
all/fnv-1a/4            time:   [225.30 ps 225.52 ps 225.76 ps]
                        thrpt:  [16.501 GiB/s 16.519 GiB/s 16.535 GiB/s]
Found 66 outliers among 1000 measurements (6.60%)
  26 (2.60%) high mild
  40 (4.00%) high severe
all/fnv-1a/16           time:   [26.060 ns 26.350 ns 26.658 ns]
                        thrpt:  [572.39 MiB/s 579.09 MiB/s 585.52 MiB/s]
Found 17 outliers among 1000 measurements (1.70%)
  14 (1.40%) high mild
  3 (0.30%) high severe
all/fnv-1a/64           time:   [175.73 ns 176.28 ns 176.93 ns]
                        thrpt:  [344.97 MiB/s 346.23 MiB/s 347.32 MiB/s]
Found 118 outliers among 1000 measurements (11.80%)
  33 (3.30%) high mild
  85 (8.50%) high severe
all/fnv-1a/256          time:   [698.70 ns 700.94 ns 703.45 ns]
                        thrpt:  [347.06 MiB/s 348.30 MiB/s 349.42 MiB/s]
Found 132 outliers among 1000 measurements (13.20%)
  27 (2.70%) high mild
  105 (10.50%) high severe
all/fnv-1a/1024         time:   [2.8155 µs 2.8266 µs 2.8387 µs]
                        thrpt:  [344.02 MiB/s 345.49 MiB/s 346.85 MiB/s]
Found 126 outliers among 1000 measurements (12.60%)
  36 (3.60%) high mild
  90 (9.00%) high severe
all/fnv-1a/4096         time:   [11.188 µs 11.223 µs 11.264 µs]
                        thrpt:  [346.80 MiB/s 348.05 MiB/s 349.13 MiB/s]
Found 55 outliers among 1000 measurements (5.50%)
  25 (2.50%) high mild
  30 (3.00%) high severe
all/fnv-1a/16384        time:   [45.074 µs 45.260 µs 45.461 µs]
                        thrpt:  [343.70 MiB/s 345.23 MiB/s 346.65 MiB/s]
Found 95 outliers among 1000 measurements (9.50%)
  32 (3.20%) high mild
  63 (6.30%) high severe
all/fnv-1a/65536        time:   [183.47 µs 184.37 µs 185.32 µs]
                        thrpt:  [337.25 MiB/s 339.00 MiB/s 340.66 MiB/s]
Found 105 outliers among 1000 measurements (10.50%)
  25 (2.50%) high mild
  80 (8.00%) high severe
all/fnv-1a/262144       time:   [725.58 µs 729.27 µs 733.40 µs]
                        thrpt:  [340.88 MiB/s 342.81 MiB/s 344.55 MiB/s]
Found 110 outliers among 1000 measurements (11.00%)
  19 (1.90%) high mild
  91 (9.10%) high severe
</pre>
</details>

**Reply by arduano** [reply]
[criterion.zip](https://github.com/ogxd/gxhash/files/13598257/criterion.zip)
Here's a zip of the full criterion folder, if needed

**Reply by ogxd** [reply]
> all/gxhash/64           time:   [169.12 ps 169.49 ps 169.90 ps]
                        thrpt:  [350.83 GiB/s 351.68 GiB/s 352.43 GiB/s]


😮 
Thanks a lot for these benches. I am not sure what is going on. Maybe there is another provider of github runners that proposes hardware on which we can observe similar results.

**Reply by arduano** [reply]
Just curious (I don't fully understand gxhash), but would it be possible to extend the <128 bytes processing functions to >=128? Just for running a benchmark here

**Reply by ogxd** [reply]
Simply put, the processing <128 bytes is a manually unrolled loop, like so:
```
if len >= 16
  hash next 16 bytes
if len >= 32
  hash next 16 bytes
if len >= 48
  hash next 16 bytes
if len >= 128
  loop by 128 bytes
     hash next 16 bytes (8 times)
  loop by 16 bytes (what remains)
     hash next 16 bytes
```
This manual unroll < 128 bytes is there to avoid the cost of jump instructions from the looping (the algorithm is so fast that the loop itself costs more than the actual block hashing on some platforms).   
The manual unroll is great however it has a cost in terms of bytecode size: because the "hash next 16 bytes" is inlined, every "if len >= X then hash next 16 bytes" makes the bytecode larger. The larger the bytecode is, the less likely it is to be inlined within functions that use the hashing algorithm or to entirely fit in the instructions cache. This is why it has to be used sparingly.  

That said, this discussion makes me think that it could be possible to get rid of the second loop, I'll run some experiments.  
Also, I may have a theory which is that the Nvidia Jetson hardware has quite a high latency on memory access, and the loops make it more difficult to prefetch. There are ways to explicitly prefetch data, and [I've seen some algorithms making use of it](https://github.com/flier/rust-t1ha/blob/45484ad71eb33d184cb5314c80f9ba6b69d17209/src/t1ha0_aes.rs#L34). I didn't see performance improvements on my laptop, so I did not include it, but maybe it would make a difference on the Nvidia Jetson.  

I'll experiment with a few ideas and get back to you. Do you think you'll still be able to run a few benches on the Jetson in a some days ?

**Reply by arduano** [reply]
Yeah just ping me, I'm on Sydney time, but I get notifications over email usually.

**Reply by ogxd** [reply]
Hello @arduano ! I was wondering how gxhash would perform on the Nvidia Jetson after all recent updates. Do you still have your hands on it?

**Reply by arduano** [reply]
Oh lol my contract ended sorry, I'm still in contact with them though and will give it a go if I get to use the Jetsons again
